src/algebra/module/basic.lean

```diff
@@ -461,19 +455,64 @@ end
 
 end add_monoid_hom
 
-section module_division_ring
-/-! Some tests for the vanishing of elements in modules over division rings. -/
+section no_zero_smul_divisors
+/-! ### `no_zero_smul_divisors`
+
+This section defines the `no_zero_smul_divisors` class, and includes some tests
+for the vanishing of elements (especially in modules over division rings).
+-/
+
+/-- `no_zero_smul_divisors R M` states that a scalar multiple is `0` only if either argument is `0`.
+
+The main application of `no_zero_smul_divisors R M`, when `M` is a semimodule,
+is the result `smul_eq_zero`: a scalar multiple is `0` iff either argument is `0`.
+
+It is a generalization of the `no_zero_divisors` class to heterogeneous multiplication.
+-/
+class no_zero_smul_divisors (R M : Type*) [has_zero R] [has_zero M] [has_scalar R M] : Prop :=
+(eq_zero_or_eq_zero_of_smul_eq_zero : ∀ {c : R} {x : M}, c • x = 0 → c = 0 ∨ x = 0)
+
+export no_zero_smul_divisors (eq_zero_or_eq_zero_of_smul_eq_zero)
```

Comment:
I think this probably belongs in `group_action.defs` so that it's near `smul_comm_class` etc.

Reply:
The reason I put it near `module` instead of `has_scalar` is that the converse (needed for `smul_eq_zero`), `x = 0 → c • x = 0`, requires `semimodule`. So if we want to generalize it, I would prefer that the field in `no_zero_smul_divisors` is an `iff`.